### ISSUE TITLE
Performance comparison

### ISSUE THREAD

**Issue by samuelcolvin** [trishume/syntect#372]
Hi, thanks for syntect it's awesome - I use it day in, day out via bat.

I've built a syntax highlight library for python using syntect (currently closed source as it's not ready to release) to replace pygments - the main reason I did this was to improve performance, however once I had the library working I found the performance was on a par with [pygments](https://github.com/pygments/pygments) which I was hoping to replace.

To make these benchmarks easier to ready, I've put them into a standalone repo: https://github.com/samuelcolvin/benchmark-highlight

TL;DR - syntect's performance is on a par with pygments which is implemented in pure python. My question is: is syntect a bit slow, or is pygments performance extremely impressive for pure python?

I guess syntact highlight is mostly running regexes, and python's `re` module has had many years of optimisation, but I'm still very surprised the python implementation has performance on a par with rust.

(By "on a par" I mean the time taken to highlight code is within 5%)

I hope I'm making some naive mistake in the rust implementation which is seriously damaging performance, but I've tried most things and most optimisation flags to no avail. Please let me know if I have?

---

Sorry there's no specific questions here, rather a query about overall performance and if there's anything I can do to improve it. Feel free to close this issue if you don't think it's reasonable.

**Comment by keith-hall** [comment]
> TL;DR - syntect's performance is on a par with pygments which is implemented in pure python. My question is: is syntect a bit slow, or is pygments performance extremely impressive for pure python?

How about a third possibility: syntect's syntax highlighting is much better quality and this is why the timings are similar. There was previous allusions to this at: https://github.com/trishume/syntect/issues/242#issuecomment-48514618

**Comment by trishume** [comment]
That seems right to me. Probably both syntect and pygments spend most of their time in the regex engine, and both use pretty similar quality C regex engines.

I would guess that pygments has simpler grammars on average, so syntect has to match more and more complex regexes against the same code than pygments, to allow highlighting more different types of syntactic constructs in different shades than pygments does. This depends on the complexity of the sublime-syntax, so you may get different relative performances for different languages, with more popular languages getting relatively worse performance in syntect (because people make the grammars fancier).

**Comment by samuelcolvin** [comment]
That makes sense. Thanks for the explanation.

Feel free to close this issue.

**Comment by sharkdp** [comment]
Slightly related: https://github.com/sharkdp/bat/blob/master/doc/alternatives.md

Note that the document hasn't been updated recently though.

**Comment by samuelcolvin** [comment]
Interesting, thank you.

**Comment by samuelcolvin** [comment]
Thanks, so in summary the fastest way to highlight code is [sourcegraph/syntaxhighlight](https://github.com/sourcegraph/syntaxhighlight) which basically "make different bits different colours and hope for the best".

I'm still surprised highlighting code can't be quicker, but I guess it would require dedicated code for each language which no-one wants to write.
